Signal/src/util/DisplayableText.swift

```diff
@@ -11,13 +11,83 @@ import Foundation
     let fullText: String
     let displayText: String
     let isTextTruncated: Bool
+    let jumbomojiCount: NSNumber?
+
+    static let kMaxJumbomojiCount: UInt = 5
 
     // MARK: Initializers
 
-    init(fullText: String, displayText: String, isTextTruncated: Bool) {
+    init(fullText: String, displayText: String, isTextTruncated: Bool, jumbomojiCount: NSNumber?) {
         self.fullText = fullText
         self.displayText = displayText
         self.isTextTruncated = isTextTruncated
+        self.jumbomojiCount = jumbomojiCount
+    }
+
+    // MARK: Emoji
+
+    private class func canDetectEmoji() -> Bool {
+        if #available(iOS 10.0, *) {
+            return true
+        } else {
+            return false
+        }
+    }
+
+    // If the string is...
+    //
+    // * Non-empty
+    // * Only contains emoji
+    // * Contains <= maxEmojiCount emoji
+    //
+    // ...return the number of emoji in the string.  Otherwise return nil.
+    //
+    // On iOS 9 and earler, always returns nil.
+    @objc public class func jumbomojiCount(to string: String) -> NSNumber? {
+        if string == "" {
+            return nil
+        }
+        if string.characters.count > Int(kMaxJumbomojiCount) {
```

Comment:
I made a few refinements to reduce perf cost, like this early exit.

Reply:
> Critically, this code assumes that one character == one emoji. I haven't found an exception yet, but I suspect there are. To be continued.

In swift String.characters represent grapheme-clusters, so this should "just work", which is nice.

> I made a few refinements to reduce perf cost, like this early exit.

The downside is , because grapheme clusters can be any length, this requires parsing the string to know how many characters it composes. Said differently, the perf of string.characters.count is related to the length of string, and potentially a bit expensive if the string is very long.

Something to consider if this proves to be a bottleneck is, instead of getting the whole count, to bail if we can enumerate beyond kMaxJumbomojiCount.

Reply:
Surely `[NSMutableAttributedString fixAttributes:]` is much more expensive than `characters.count`, since it processes the whole string.  That's the work I'm trying to avoid here.  Having said that, `fixAttributes:` takes a range - perhaps we could skip the string length check and pass a shorter range to `fixAttributes:`.  That feels like over-optimizing though at the cost of readability.  

Bear in mind that we're now doing a much better job avoiding this work.

* `ConversationViewItem` does this work lazily.
* `ConversationViewItem` never does this work twice, per instance of `ConversationViewController`, since the result is expensive to compute and very cheap to cache.  

Therefore I really doubt this is a hotspot, so long as we do the string length check.